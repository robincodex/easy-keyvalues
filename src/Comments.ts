export class KeyValuesComments {
    constructor(
        protected comments: string[] = [],
        protected endOfLineComment: string = '',
    ) {}

    public GetComments(): string[] {
        return this.comments;
    }

    public HasComments() {
        return this.comments.length > 0;
    }

    public SetComments(list: string[]) {
        for (const v of list) {
            if (v.includes('\n')) {
                throw Error('The comment only allowed one line');
            }
        }
        this.comments = list;
        return this;
    }

    public AppendComment(text: string) {
        if (text.includes('\n')) {
            throw Error('The comment only allowed one line');
        }
        this.comments.push(text);
        return this;
    }

    public SetEndOfLineComment(text: string) {
        if (text.includes('\n')) {
            throw Error('The comment only allowed one line');
        }
        this.endOfLineComment = text;
        return this;
    }

    public GetEndOfLineComment() {
        return this.endOfLineComment;
    }

    public HasEndOfLineComment() {
        return this.endOfLineComment.length > 0;
    }
}

export class KeyValues3Comments {
    constructor(
        protected comments: string[] = [],
        protected endOfLineComment: string = '',
    ) {}

    public GetComments(): string[] {
        return this.comments;
    }

    public HasComments() {
        return this.comments.length > 0;
    }

    public SetComments(list: string[]) {
        this.comments = list;
        return this;
    }

    public AppendComment(text: string) {
        this.comments.push(text);
        return this;
    }

    public SetEndOfLineComment(text: string) {
        if (text.includes('\n')) {
            throw Error('The end of line comment only allowed one line');
        }
        this.endOfLineComment = text;
        return this;
    }

    public GetEndOfLineComment() {
        return this.endOfLineComment;
    }

    public HasEndOfLineComment() {
        return this.endOfLineComment.length > 0;
    }

    public Format(tab: string = ''): string {
        let text = '';
        for (const comment of this.comments) {
            if (comment.includes('\n')) {
                const lines = comment.split('\n').map((v) => v.trimStart());
                text += `${tab}/*\n`;
                if (lines.some((v) => v.startsWith('*'))) {
                    text += lines
                        .map((v) => {
                            if (v.startsWith('*')) {
                                v = v.trimStart();
                            } else {
                                v = '* ' + v.trimStart();
                            }
                            return `${tab} ` + v + '\n';
                        })
                        .join('');
                    text += `${tab} */\n`;
                } else {
                    text += lines.map((v) => `${tab}` + v + '\n').join('');
                    text += `${tab}*/\n`;
                }
            } else {
                text += `${tab}// ${comment.trimStart()}\n`;
            }
        }
        return text;
    }
}
