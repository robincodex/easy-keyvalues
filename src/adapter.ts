export interface KeyValuesAdapter {
    readFile(path: string, encoding?: string): Promise<string>;
    writeFile(path: string, data: string, encoding?: string): Promise<void>;
    resolvePath(filename: string, basePath: string): string;
    createKeyValuesID(): string;
}

let defaultAdapter: KeyValuesAdapter = {
    async readFile() {
        return '';
    },
    async writeFile() {},
    resolvePath(filename: string, basePath: string) {
        return filename + basePath;
    },
    createKeyValuesID() {
        return '';
    },
};

export function setKeyValuesAdapter(adapter: KeyValuesAdapter): void {
    defaultAdapter = adapter;
}

export function getKeyValuesAdapter(): KeyValuesAdapter {
    return defaultAdapter;
}
