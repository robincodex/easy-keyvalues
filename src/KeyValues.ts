import { getKeyValuesAdapter } from './adapter';
import { KeyValuesComments } from './Comments';

const KeyValuesRootKey = '__KeyValues_Root__';

function createID() {
    const adapter = getKeyValuesAdapter();
    return adapter.createKeyValuesID();
}

export default class KeyValues {
    private __filename?: string;

    public get filename() {
        return this.__filename || '';
    }

    public set filename(s: string | undefined) {
        this.__filename = s ? s.replace(/\\/g, '/') : s;
    }

    /**
     * The children of this KeyValues
     */
    protected children?: KeyValues[];
    /**
     * The value of this KeyValues
     */
    protected value?: string;
    /**
     * The parent of this KeyValues,
     */
    protected parent?: KeyValues;
    /**
     * Comment
     */
    public Comments = new KeyValuesComments();
    /**
     * The KeyValues flags, such as [$WIN32] [$X360]
     */
    public Flags: string = '';
    /**
     * Unique id of KeyValues
     */
    public readonly ID = createID();

    constructor(
        public Key: string,
        defaultValue?: string | KeyValues[],
    ) {
        this.SetValue(defaultValue || '');
    }

    /**
     * The parent of this KeyValues
     */
    public GetParent() {
        return this.parent;
    }

    /**
     * Return true that the KeyValues is root.
     */
    public IsRoot() {
        return this.Key === KeyValuesRootKey;
    }

    /**
     * The key is #base?
     */
    public IsBase() {
        return this.Key === '#base';
    }

    protected baseFilePath = '';

    /**
     * Return #base's value
     */
    public GetBaseFilePath() {
        return this.baseFilePath;
    }

    /**
     * KeyValues list of #base
     */
    public GetBaseList() {
        return this.FindAllKeys('#base');
    }

    /**
     * The children of this KeyValues,
     * if no children then return empty array.
     */
    public GetChildren(): Readonly<KeyValues[]> {
        return this.children || [];
    }

    public GetChildCount() {
        return this.GetChildren().length;
    }

    public GetFirstChild(): KeyValues | undefined {
        return this.GetChildren()[0];
    }

    public GetLastChild(): KeyValues | undefined {
        return this.GetChildren()[this.GetChildCount() - 1];
    }

    /**
     * Create a KeyValues to children and return it.
     */
    public CreateChild(key: string, value: string | KeyValues[]) {
        const kv = new KeyValues(key, value);
        this.Append(kv);
        return kv;
    }

    /**
     * The value of this KeyValues,
     * if no value then return empty string.
     */
    public GetValue() {
        return this.value || '';
    }

    /**
     * Return true that the KeyValues exists children and no value.
     */
    public HasChildren() {
        return !!this.children;
    }

    /**
     * Set value or children.
     */
    public SetValue(v: string | KeyValues[]) {
        if (Array.isArray(v)) {
            this.children = v.map((c) => c.Free());
            delete this.value;
            for (const kv of this.children) {
                if (kv === this) {
                    throw new Error(`SetValue(): The value can not includes self`);
                }
                kv.parent = this;
            }
        } else {
            if (this.IsRoot()) {
                throw new Error('The value of the root node kv must be an array');
            }
            this.value = v;
            delete this.children;
        }
        return this;
    }

    /**
     * Append a KeyValues to children,
     * if no children then throw error.
     */
    public Append(child: KeyValues) {
        if (this.children) {
            if (child === this) {
                throw new Error(`Append(): Can not append self`);
            }
            this.children.push(child.Free());
            child.parent = this;
        } else {
            throw new Error(`The KeyValues [Key = ${this.Key}] does not have children`);
        }
        return this;
    }

    /**
     * Insert a KeyValues to children,
     * if no children then throw error.
     */
    public Insert(child: KeyValues, index: number) {
        if (this.children) {
            if (child === this) {
                throw new Error(`Insert(): Can not insert self`);
            }
            this.children.splice(index, 0, child.Free());
            child.parent = this;
        } else {
            throw new Error(`The KeyValues [Key = ${this.Key}] does not have children`);
        }
        return this;
    }

    /**
     * Find a KeyValues from children
     */
    public Find(
        callback: (kv: KeyValues, i: number, parent: KeyValues) => boolean,
    ): KeyValues | undefined {
        if (!this.children) {
            return;
        }
        for (const [i, kv] of this.children.entries()) {
            if (callback(kv, i, this) === true) {
                return kv;
            }
        }
    }

    /**
     * Find all KeyValues from children
     */
    public FindAll(
        callback: (kv: KeyValues, i: number, parent: KeyValues) => boolean,
    ): KeyValues[] {
        if (!this.children) {
            return [];
        }
        const result: KeyValues[] = [];
        for (const [i, kv] of this.children.entries()) {
            if (callback(kv, i, this) === true) {
                result.push(kv);
            }
        }
        return result;
    }

    /**
     * Find a KeyValues
     */
    public FindKey(key: string) {
        return this.Find((kv) => kv.Key === key);
    }

    /**
     * Find all KeyValues
     */
    public FindAllKeys(...keys: string[]) {
        return this.FindAll((kv) => keys.includes(kv.Key));
    }

    /**
     * Find a KeyValues from children and children's children...
     */
    public FindTraverse(
        callback: (kv: KeyValues, i: number, parent: KeyValues) => boolean,
    ): KeyValues | undefined {
        if (!this.children) {
            return;
        }
        return KeyValues.FindTraverse(this, callback);
    }

    /**
     * Find a KeyValues from children and children's children...
     */
    protected static FindTraverse(
        root: KeyValues,
        callback: (kv: KeyValues, i: number, parent: KeyValues) => boolean,
    ): KeyValues | undefined {
        if (root.HasChildren()) {
            for (const [i, kv] of root.GetChildren().entries()) {
                if (callback(kv, i, root) === true) {
                    return kv;
                }
                const result = KeyValues.FindTraverse(kv, callback);
                if (result) {
                    return result;
                }
            }
        }
    }

    /**
     * Find child from the current KeyValues
     */
    public FindID(id: string): KeyValues | undefined {
        return this.Find((kv) => kv.ID === id);
    }

    /**
     * Recursively iterate through all children to find the value that matches the ID
     */
    public FindIDTraverse(id: string): KeyValues | undefined {
        return this.FindTraverse((kv) => kv.ID === id);
    }

    /**
     * Delete a KeyValues from children
     */
    public Delete(child: string | KeyValues): KeyValues | undefined {
        if (!this.children) {
            return;
        }
        let kv: KeyValues | undefined;
        if (typeof child === 'string') {
            kv = this.children.find((v) => v.Key === child);
        } else {
            kv = this.children.find((v) => v === child);
        }
        if (kv) {
            this.children = this.children.filter((v) => v !== kv);
            kv.Free();
        }
        return kv;
    }

    /**
     * Delete this KeyValues from parent
     */
    public Free() {
        if (this.parent) {
            this.parent.Delete(this);
            delete this.parent;
        }
        return this;
    }

    /**
     * Format KeyValues to file text
     */
    public Format(tab: string = '', maxLength: number = -1): string {
        if (this.IsRoot()) {
            if (!this.children) {
                throw new Error('The value of the root node kv must be an array');
            }
            return this.children.map((v) => v.Format()).join('\n');
        }

        let text = '';

        if (this.Comments.HasComments()) {
            text += this.Comments.GetComments()
                .map((v) => `${tab}// ${v.trimStart()}\n`)
                .join('');
        }

        if (this.children) {
            if (this.IsBase()) {
                text += `${tab}${this.Key}    "${this.GetBaseFilePath()}"`;
                if (this.Comments.HasEndOfLineComment()) {
                    text += ` // ${this.Comments.GetEndOfLineComment()}`;
                }
                return text;
            }
            const maxLength = Math.max(...this.children.map((v) => v.Key.length));
            text += `${tab}"${this.Key}"`;
            if (this.Comments.HasEndOfLineComment()) {
                text += ` // ${this.Comments.GetEndOfLineComment()}`;
            }
            text += `\n${tab}{`;
            for (const kv of this.children) {
                text += '\n' + kv.Format(tab + '    ', maxLength);
            }
            text += `\n${tab}}`;
        } else {
            if (this.IsBase()) {
                text += `${tab}${this.Key}    "${this.value}"`;
                if (this.Comments.HasEndOfLineComment()) {
                    text += ` // ${this.Comments.GetEndOfLineComment()}`;
                }
            } else {
                text += `${tab}"${this.Key}"${' '.repeat(
                    Math.max(0, maxLength - this.Key.length),
                )}`;
                text += `    "${this.value}"`;
                if (this.Comments.HasEndOfLineComment()) {
                    text += ` // ${this.Comments.GetEndOfLineComment()}`;
                }
            }
        }
        return text;
    }

    public toString() {
        return this.Format();
    }

    /**
     * Deep clone KeyValues
     */
    public Clone(): KeyValues {
        if (!this.children) {
            return new KeyValues(this.Key, this.value);
        }
        return new KeyValues(
            this.Key,
            this.children.map((v) => v.Clone()),
        );
    }

    /**
     * Create root node
     */
    public static CreateRoot() {
        return new KeyValues(KeyValuesRootKey, []);
    }

    /**
     * Parse string
     */
    public static async Parse(body: string, filename?: string): Promise<KeyValues> {
        let root = this.CreateRoot();
        root.filename = filename;
        this._parse({ body, pos: 0, line: 1 }, root);

        // Load #base
        if (filename) {
            const baseList = root.FindAllKeys('#base');
            const adapter = getKeyValuesAdapter();
            for (const base of baseList) {
                const v = base.GetValue().trim();
                if (v) {
                    base.baseFilePath = v;
                    const filePath = adapter.resolvePath(filename, v);
                    const baseKV = await KeyValues.Load(filePath);
                    base.filename = baseKV.filename;
                    base.SetValue(baseKV.GetChildren().map((v) => v.Free()));
                }
            }
        }

        return root;
    }

    protected static _parse(data: { body: string; pos: number; line: number }, parent: KeyValues) {
        let kv = new KeyValues('');
        let key = false;
        let leftMark = false;
        let inQoute = false;
        let str = '';
        let isEndOfLineComment = false;
        let lastKV: KeyValues | undefined;
        let matchFlag = false;
        for (; data.pos < data.body.length; data.pos++) {
            const c = data.body[data.pos];
            const isNewLine = c === '\n';
            const isSpace = isNewLine || c === ' ' || c === '\t' || c === '\r';

            if (isNewLine) {
                data.line += 1;
                isEndOfLineComment = false;
            }

            // Merge flags text
            if (lastKV && matchFlag) {
                if (c === ']') {
                    lastKV.Flags = str;
                    matchFlag = false;
                    str = '';
                    continue;
                }
                str += c;
                continue;
            }

            // If leftMark is true then merge char to str
            if (leftMark) {
                if (!inQoute) {
                    if (c === '{' || c === '"' || c === '[' || c === ']') {
                        throw Error(`Not readable in line ${data.line}`);
                    }
                    if (isSpace || c === '}') {
                        if (key) {
                            kv.Key = str;
                        } else {
                            kv.SetValue(str);
                            parent.Append(kv);
                            lastKV = kv;
                            kv = new KeyValues('');
                        }
                        leftMark = false;
                        inQoute = false;
                        str = '';
                        if (c === '}') {
                            data.pos--;
                        }
                        continue;
                    }
                    str += c;
                    continue;
                }
                if (c === '\\') {
                    data.pos++;
                    str += c + data.body[data.pos];
                    continue;
                }
                if (c === '"') {
                    if (key) {
                        kv.Key = str;
                    } else {
                        kv.SetValue(str);
                        parent.Append(kv);
                        lastKV = kv;
                        kv = new KeyValues('');
                    }
                    leftMark = false;
                    inQoute = false;
                    str = '';
                    continue;
                }
                str += c;
                continue;
            }

            // if comment
            if (c === '/') {
                if (data.body[data.pos + 1] === '*') {
                    throw Error(`Line ${data.line}: not support multi-line comment`);
                }
                if (data.body[data.pos + 1] === '/') {
                    const endIndex = data.body.indexOf('\n', data.pos + 1);
                    const comment = data.body.slice(data.pos + 2, endIndex).trim();
                    if (comment) {
                        if (isEndOfLineComment) {
                            if (key) {
                                kv.Comments.SetEndOfLineComment(comment);
                            } else {
                                const lastChild = parent.GetLastChild();
                                if (lastChild) {
                                    lastChild.Comments.SetEndOfLineComment(comment);
                                }
                            }
                        } else {
                            kv.Comments.AppendComment(comment);
                        }
                    }
                    isEndOfLineComment = false;
                    data.pos = endIndex;
                    continue;
                }
            }

            // If open breace
            if (c === '{') {
                data.pos++;
                kv.SetValue([]);
                parent.Append(kv);
                this._parse(data, kv);
                lastKV = kv;
                kv = new KeyValues('');
                key = false;
                continue;
            }

            // If close breace
            if (c === '}') {
                data.pos++;
                break;
            }

            // If space
            if (isSpace) {
                continue;
            }

            // Match flag
            if (lastKV && c === '[') {
                str = '';
                matchFlag = true;
                continue;
            }

            // start merge char
            key = !key;
            leftMark = true;
            inQoute = c === '"';
            str = inQoute ? '' : c;
            isEndOfLineComment = true;
        }
    }

    /**
     * Convert KeyValues to object and exclude comments.
     */
    public toObject<T = any>(): T {
        const obj: any = {};
        if (!this.HasChildren()) {
            throw Error('Not found children in this KeyValues');
        }

        for (const kv of this.children!) {
            if (!kv.HasChildren()) {
                obj[kv.Key] = kv.GetValue();
            } else {
                obj[kv.Key] = kv.toObject();
            }
        }

        return obj;
    }

    /**
     * Load KeyValues from file
     */
    public static async Load(filename: string, encoding?: string): Promise<KeyValues> {
        const adapter = getKeyValuesAdapter();
        const text = await adapter.readFile(filename, encoding);
        if (text.charCodeAt(0) === 0xfeff) {
            return await this.Parse(text.slice(1), filename);
        }
        return await this.Parse(text, filename);
    }

    /**
     * Save KeyValues to file
     */
    public async Save(otherFilename?: string, encoding?: string): Promise<void> {
        const filename = otherFilename ?? this.filename;
        if (!filename) {
            throw new Error('Not found filename in KeyValues');
        }
        const adapter = getKeyValuesAdapter();
        await adapter.writeFile(filename, this.Format(), encoding);

        // Save #base
        const baseList = this.FindAllKeys('#base');
        for (const base of baseList) {
            const content = base
                .GetChildren()
                .map((v) => v.Format())
                .join('\n');
            if (otherFilename) {
                const filePath = adapter.resolvePath(filename, base.GetBaseFilePath());
                await adapter.writeFile(filePath, content, encoding);
            } else {
                await adapter.writeFile(base.filename!, content, encoding);
            }
        }
    }
}
