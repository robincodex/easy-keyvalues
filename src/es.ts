import KeyValues from './KeyValues';
import KeyValues3 from './KeyValues3';
import { KeyValuesComments, KeyValues3Comments } from './Comments';

export { KeyValues, KeyValues3, KeyValuesComments, KeyValues3Comments };

export * from './adapter';
