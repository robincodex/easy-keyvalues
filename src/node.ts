import { promises } from 'fs';
import { resolve } from 'path';
import { KeyValues3Comments, KeyValuesComments } from './Comments';
import KeyValues from './KeyValues';
import KeyValues3 from './KeyValues3';
import { setKeyValuesAdapter } from './adapter';

const { readFile, writeFile } = promises;

export * from './adapter';

export { KeyValues, KeyValues3, KeyValues3Comments, KeyValuesComments };

setKeyValuesAdapter({
    async readFile(path, encoding = 'utf8') {
        const buf = await readFile(path, encoding as BufferEncoding);
        return buf.toString();
    },
    async writeFile(path, data, encoding = 'utf8') {
        await writeFile(path, data, encoding as BufferEncoding);
    },
    resolvePath(filename, basePath) {
        return resolve(filename, '../' + basePath);
    },
    createKeyValuesID() {
        return '';
    },
});
