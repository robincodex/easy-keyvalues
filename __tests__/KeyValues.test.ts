import { describe, expect, test } from '@jest/globals';
import crypto from 'crypto';
import { join } from 'path';
import { KeyValues, getKeyValuesAdapter } from '../src/node';

function testKV(kv: KeyValues) {
    expect(kv.GetChildCount()).toBe(3);

    expect(kv.GetChildren().map((v) => v.Key)).toEqual(['DOTAAbilities', 'Ha', 'test']);

    expect(kv.Find((kv) => kv.Key === 'Ha')?.Key).toBe('Ha');
    expect(kv.Find((kv) => kv.Key === '')?.Key).toBe(undefined);
    expect(kv.FindTraverse((kv) => kv.Key === '')?.Key).toBe(undefined);

    const findAll = kv.FindKey('Ha')?.FindAll((kv) => kv.Key === 'test');
    expect(findAll?.length).toBe(3);
    expect(findAll?.map((v) => v.Key)).toEqual(['test', 'test', 'test']);
    expect(findAll?.map((v) => v.GetValue())).toEqual(['123', '345', '789']);

    expect(kv.FindTraverse((kv) => kv.Key === 'BaseClass')?.GetValue()).toBe('ability_datadriven');

    expect(kv.FindTraverse((kv) => kv.Key === 'k3')?.GetValue()).toBe('v3');

    expect(kv.FindKey('DOTAAbilities')?.FindKey('BaseClass')?.GetValue()).toBe('ability_lua');

    expect(
        kv
            .FindKey('DOTAAbilities')
            ?.FindKey('building_system_active_preview')
            ?.FindKey('AbilityTextureName')
            ?.GetValue(),
    ).toBe('abaddon/mistral_fiend_icons//abaddon_borrowed_time');

    expect(kv.FindKey('test')?.FindKey('children')?.FindKey('c2')?.GetValue()).toBe('2 \\"two\\"');

    expect(kv.FindKey('test')?.Comments?.GetComments()[0]).toBe('test object');

    expect(kv.FindKey('test')?.FindKey('children')?.Comments?.GetComments()).toEqual([
        'comment02',
        'comment03',
        'comment04',
    ]);

    expect(kv.FindKey('test')?.FindKey('children')?.Comments?.GetEndOfLineComment()).toBe(
        'this is children',
    );

    expect(kv.FindKey('test')?.FindKey('children')?.FindKey('c3')?.GetValue()).toBe(
        'sss\n    <br/>this is child\\n index 3\n    88',
    );
}

describe('KeyValues', () => {
    test('Check KeyValues.txt', async () => {
        const kv = await KeyValues.Load(join(__dirname, 'KeyValues.txt'));
        testKV(kv);
        expect(kv.toString()).toMatchSnapshot();
    });

    test('Check chat_english.txt', async () => {
        const kv = await KeyValues.Load(join(__dirname, 'chat_english.txt'));
        const tokens = kv.FindKey('lang')?.FindKey('Tokens');
        expect(!tokens).toBe(false);
        expect(tokens?.GetChildCount()).toBe(11);
    });

    test('Check gameui_english.txt', async () => {
        const kv = await KeyValues.Load(join(__dirname, 'gameui_english.txt'));
        const tokens = kv.FindKey('lang')?.FindKey('Tokens');
        expect(!tokens).toBe(false);
        expect(tokens?.GetChildCount()).toBe(688);
        expect(tokens?.FindKey('GameUI_JoystickMoveLookSticks')?.Flags).toBe('$WIN32');
        expect(
            tokens
                ?.FindAllKeys('GameUI_JoystickMoveLookSticks')
                ?.map((v) => v.Flags)
                .join(','),
        ).toBe('$WIN32,$X360');
    });

    test('Check KeyValues.save.txt', async () => {
        const kv = await KeyValues.Load(join(__dirname, 'KeyValues.save.txt'));
        testKV(kv);
        await kv.Save();
    });

    test('Check KeyValues Parse Error', async () => {
        await KeyValues.Parse(`a 123\nb {c 123}`);
        expect(async () => {
            await KeyValues.Parse(`a{} 123`);
        }).rejects.toThrow();

        expect(async () => {
            await KeyValues.Parse(`a[] 123`);
        }).rejects.toThrow();

        expect(async () => {
            await KeyValues.Parse(`a" 123`);
        }).rejects.toThrow();

        expect(async () => {
            await KeyValues.Parse(`/* a */`);
        }).rejects.toThrow();
    });

    test('Create/Append/Insert/Delete KeyValues', async () => {
        const root = KeyValues.CreateRoot();
        root.CreateChild('a', 'b');
        root.CreateChild('#base', 'file.kv');
        expect(root.toString()).toBe(`"a"    "b"\n#base    "file.kv"`);
        expect(root.GetParent()).toBe(undefined);
        expect(root.GetFirstChild()?.Key).toBe('a');
        expect(root.GetFirstChild()?.GetParent() === root).toBe(true);

        root.CreateChild('c', [new KeyValues('d', 'a')])
            .Append(new KeyValues('d', 'a'))
            .Insert(new KeyValues('d', [new KeyValues('d', 'd')]), 0);
        expect(root.toString()).toBe(
            `"a"    "b"
#base    "file.kv"
"c"
{
    "d"
    {
        "d"    "d"
    }
    "d"    "a"
    "d"    "a"
}`,
        );

        const cloneRoot = root.Clone();
        expect(cloneRoot === root).toBe(false);
        expect(cloneRoot.toString()).toBe(root.toString());

        cloneRoot.GetFirstChild()!.Key = 'b';
        cloneRoot.FindKey('c')?.FindKey('d')?.Append(new KeyValues('c', 'c'));
        expect(root.GetFirstChild()!.Key).toBe('a');
        expect(cloneRoot.FindKey('c')?.FindKey('d')?.FindKey('c')?.GetValue()).toBe('c');
        expect(root.FindKey('c')?.FindKey('d')?.FindKey('c') === undefined).toBe(true);

        const c = root.Delete('c');
        expect(c?.toString()).toBe(`"c"
{
    "d"
    {
        "d"    "d"
    }
    "d"    "a"
    "d"    "a"
}`);
        expect(c?.GetParent() === undefined).toBe(true);
    });

    test('Check Comment', async () => {
        const testComment = new KeyValues('d', 'a');
        testComment.Comments.SetComments(['line1', 'line2']).SetEndOfLineComment('end');
        expect(testComment.toString()).toBe(`// line1
// line2
"d"    "a" // end`);

        expect(() => {
            testComment.Comments.AppendComment('a\nc');
        }).toThrow();

        expect(() => {
            testComment.Comments.SetEndOfLineComment('a\nc');
        }).toThrow();

        expect(() => {
            testComment.Comments.SetComments(['a\nc']);
        }).toThrow();
    });

    test('Check KeyValues Error', () => {
        const kv = new KeyValues('a', []);
        expect(kv.GetValue()).toBe('');
        kv.SetValue('');
        expect(kv.GetChildren().length).toBe(0);

        expect(() => {
            kv.Append(new KeyValues(''));
        }).toThrow();

        expect(() => {
            kv.Insert(new KeyValues(''), 0);
        }).toThrow();

        expect(async () => {
            await kv.Save();
        }).rejects.toThrow();

        expect(kv.Find((v) => true) === undefined).toBe(true);
        expect(kv.FindKey('') === undefined).toBe(true);
        expect(kv.FindTraverse((v) => true) === undefined).toBe(true);
        expect(kv.FindAll((v) => true).length).toEqual(0);
        expect(kv.FindAllKeys('').length).toEqual(0);
        expect(kv.Delete('') === undefined).toBe(true);

        expect(() => {
            const root = KeyValues.CreateRoot();
            // @ts-ignore
            root.value = '';
            // @ts-ignore
            delete root.children;
            root.Format();
        }).toThrow();

        expect(() => {
            KeyValues.CreateRoot().SetValue('');
        }).toThrow();

        expect(() => {
            const kv = new KeyValues('a', []);
            kv.Append(kv);
        }).toThrow();

        expect(() => {
            const kv = new KeyValues('a', []);
            kv.Insert(kv, 0);
        }).toThrow();

        expect(() => {
            const kv = new KeyValues('a', []);
            kv.SetValue([kv]);
        }).toThrow();
    });

    test('Check KeyValues #base', async () => {
        const root = await KeyValues.Load(join(__dirname, 'KeyValues.base.txt'));
        expect(
            root.FindKey('DOTAAbilities')?.FindKey('ability01')?.FindKey('BaseClass')?.GetValue(),
        ).toBe('ability_datadriven');
        expect(
            root.FindKey('DOTAAbilities')?.FindKey('ability02')?.FindKey('BaseClass')?.GetValue(),
        ).toBe(undefined);

        const root2 = await KeyValues.Load(join(__dirname, 'KeyValues.base.txt'));
        const list = root2.GetBaseList();
        expect(list[0].GetBaseFilePath()).toBe('npc/file01.txt');
        expect(list[1].GetBaseFilePath()).toBe('npc/file02.txt');
        expect(list[0].filename).toBe(join(__dirname, 'npc/file01.txt').replace(/\\/g, '/'));
        expect(list[0].toString()).toBe(`#base    "npc/file01.txt" // file01`);
        expect(list[1].toString()).toBe(`#base    "npc/file02.txt" // file02`);

        await root.Save();
        await root.Save(join(__dirname, 'base_other/root.txt'));

        const base = new KeyValues('#base', 'npc/file01.txt');
        base.Comments.SetEndOfLineComment('file01');
        expect(base.toString()).toMatchSnapshot();
    });

    test('Check KeyValues.toObject', async () => {
        const kv = await KeyValues.Parse(`a 123\nb {c 123}`);
        const obj = kv.toObject<{ a: string; b: { c: string } }>();
        expect(obj['a']).toBe('123');
        expect(obj['b']['c']).toBe('123');

        const kv2 = await KeyValues.Load(join(__dirname, 'KeyValues.txt'));
        const obj2 = kv2.toObject();
        expect(obj2['DOTAAbilities']['building_system_active_preview']['MaxLevel']).toBe('1');
        expect(typeof obj2['Ha']['t1']['ggg']).toBe('object');
        expect(typeof obj2['test']).toBe('object');

        try {
            new KeyValues('a', '123').toObject();
        } catch (e) {
            expect(e).toEqual(Error(`Not found children in this KeyValues`));
        }
    });

    test('Check KeyValues.ID', async () => {
        const noRootID = KeyValues.CreateRoot();
        expect(noRootID.ID).toBe('');

        getKeyValuesAdapter().createKeyValuesID = () => {
            return crypto.randomBytes(8).toString('hex');
        };
        const root = KeyValues.CreateRoot();
        expect(root.ID).toHaveLength(16);

        const a = new KeyValues('a', 'a');
        const b = new KeyValues('b', 'b');
        const d = new KeyValues('d', 'd');
        const e = new KeyValues('e', [d]);
        const c = new KeyValues('a', [e]);
        root.Append(a);
        root.Append(b);
        root.Append(c);
        expect(root.FindID(a.ID) === a).toBe(true);
        expect(root.FindID(b.ID) === b).toBe(true);
        expect(root.FindID(c.ID) === c).toBe(true);
        expect(root.FindID(d.ID) === undefined).toBe(true);
        expect(root.FindID(e.ID) === undefined).toBe(true);
        expect(root.FindIDTraverse(a.ID) === a).toBe(true);
        expect(root.FindIDTraverse(b.ID) === b).toBe(true);
        expect(root.FindIDTraverse(c.ID) === c).toBe(true);
        expect(root.FindIDTraverse(d.ID) === d).toBe(true);
        expect(root.FindIDTraverse(e.ID) === e).toBe(true);
    });
});
