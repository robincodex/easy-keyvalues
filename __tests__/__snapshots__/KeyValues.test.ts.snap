// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyValues Check KeyValues #base 1`] = `"#base    "npc/file01.txt" // file01"`;

exports[`KeyValues Check KeyValues.toObject 1`] = `
{
  "DOTAAbilities": {
    "ability01": {
      "AbilityBehavior": "DOTA_ABILITY_BEHAVIOR_POINT",
      "BaseClass": "ability_lua",
    },
    "ability02": {
      "AbilityBehavior": "DOTA_ABILITY_BEHAVIOR_POINT",
      "BaseClass": "ability_datadriven",
    },
  },
}
`;

exports[`KeyValues Check KeyValues.txt 1`] = `
""DOTAAbilities"
{
    "building_system_active_preview"
    {
        "BaseClass"             "ability_datadriven"
        "AbilityBehavior"       "DOTA_ABILITY_BEHAVIOR_POINT"
        "AbilityTextureName"    "abaddon/mistral_fiend_icons//abaddon_borrowed_time"
        "MaxLevel"              "1"
        "c1"
        {
            "kkkkkkkkkkkkkkkk1"    "v1"
            "kkkkkkkkkkkkkk2"      "v2"
            "kkkkkkkk3"            "v3"
            "kkkkkkkk3"            "v3"
            "kkkkkkkk3"            "v3 \\"test\\""
            "c2"
            {
                "kkkkkkkkkkk1"    "v1"
                "kkkkkkkk2"       "v2"
                "kkkkk3"          "v3"
                "kkkkkkkk3"       "v3"
                "kkkkkkkk3"       "v3"
                "kkkkkkkk3"       "v3"
                "c3"
                {
                    "kkkkkkk1"     "v1"
                    "kkkkk2"       "v2"
                    "kkkkk3"       "v3"
                    "kkkkkkkk3"    "v3"
                    "kkkkkkkk3"    "v3" // haha
                    // list
                    "c4"
                    {
                        "kkkkkk1"      "v1"
                        "kkkkk2"       "v2"
                        "kkkk3"        "v3"
                        "kkkkkkkk3"    "v3"
                        "c5"
                        {
                            "kkkk1"    "v1"
                            "kkkk2"    "v2"
                            "k3"       "v3"
                        }
                    }
                }
            }
        }
    }
    "BaseClass"                         "ability_lua"
    "AbilityBehavior"                   "DOTA_ABILITY_BEHAVIOR_POINT"
    "AbilityTextureName"                "sylph_flash_blast"
    "MaxLevel"                          "1"
}
// -> ha
"Ha"
{
    "test"      "123"
    "01haha"    "false"
    "t1"
    {
        "asd"    "456"
        "ggg"
        {
            "hjjj"    "sss"
        }
    }
    "test"      "345"
    "test"      "789"
}
// test object
"test"
{
    // comment01
    "key"         "value"
    // comment02
    // comment03
    // comment04
    "children" // this is children
    {
        "c1"    "1" // comment05
        "c2"    "2 \\"two\\""
        // comment07
        // comment08
        "c3"    "sss
    <br/>this is child\\n index 3
    88"
        "c4"
        {
            "one"    "two"
        }
    }
    "test"
    {
        "three"    "two"
    }
}"
`;
